Add `cupy.select`

niteya-shah opened the pull request:
This PR implements the function cupy.select

emcastillo commented on cupy/indexing/indexing.py:
```diff
@@ -164,4 +164,60 @@ def extract(condition, a):
+
+    dtype = cupy.result_type(*choicelist)
+
+    choicelist = [cupy.asarray(choice) for choice in choicelist]
```
cupy functions only accepts cupy arrays, so instead of `asarray` you should raise a `TypeError` for different types

niteya-shah replied:
Done

emcastillo commented on cupy/indexing/indexing.py:
```diff
@@ -164,4 +164,60 @@ def extract(condition, a):
+    dtype = cupy.result_type(*choicelist)
+
+    choicelist = [cupy.asarray(choice) for choice in choicelist]
+    choicelist.append(cupy.asarray(default))
```
ditto

niteya-shah replied:
Done

emcastillo commented:
Can you compare the time for several array sizes using `cupyx.time.repeat` and paste the outputs here?, the larger the arrays the better.

niteya-shah commented:
This is the small test code that I wrote.
```
import cupy
import cupyx

size = 10000000

arr1 = cupy.random.rand(size)
arr2 = cupy.random.rand(size)
arr3 = cupy.random.rand(size)
arr4 = cupy.random.rand(size)

selection1 = cupy.random.rand(size) < 0.5
selection2 = cupy.random.rand(size) > 0.5
selection3 = cupy.random.rand(size) < 0.1
selection4 = cupy.random.rand(size) > 0.1

args = ([selection1, selection2, selection3, selection4],
        [arr1, arr2, arr3, arr4], 1)

result = cupyx.time.repeat(cupy.select, args)
print(result)

```
For that I got 
> select              :   173.476 us   +/-112.873 (min:   72.751 / max: 1746.157) us   6462.655 us   +/-193.038 (min: 6376.800 / max: 9839.808) us

its seems that CPU processing is significantly faster than GPU for this function, so how should I proceed with this ?

emcastillo commented:
That is expected, can you post numpy times too?

niteya-shah commented:
> select              : 224782.817 us   +/-20326.572 (min:181878.700 / max:368489.397) us  224811.465 us   +/-20328.655 (min:181898.148 / max:368517.639) us

emcastillo commented on cupy/indexing/indexing.py:
```diff
@@ -164,4 +164,60 @@ def extract(condition, a):
+    choicelist = choicelist[-2::-1]
+    condlist = condlist[::-1]
+    for choice, cond in zip(choicelist, condlist):
+        cupy.copyto(result, choice, where=cond)
```
I think that instead of the multiple copy to and this loop, it should be possible to create two single cupy arrays with all the choice lists and condlists and do the selection in one single call.

emcastillo commented on tests/cupy_tests/indexing_tests/test_indexing.py:
```diff
@@ -224,3 +226,65 @@ def test_choose_broadcast_fail(self, xp, dtype):
+@testing.gpu
+class TestSelect(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_bool=True, no_complex=True)
```
Please add tests for invalid parameters, empty lists and different sized list.

niteya-shah replied:
Done

niteya-shah commented:
@emcastillo I removed the for loop and replaced them with argmax and choose, but the problem that arises because of this is that the time of execution increases drastically.

> size = 100000
> numpy
> select : 1964.875 us +/-276.154 (min: 1743.394 / max: 2950.273) us
> cupy
> select : 568.441 us +/-14.084 (min: 555.589 / max: 633.240) us 604.920 us +/-14.027 (min: 593.312 / max: 669.504) us
> 

> size = 1000000
> numpy
> select : 18466.939 us +/-996.490 (min:17225.177 / max:21887.793) us
> cupy
> select : 4761.512 us +/-402.079 (min: 4310.876 / max: 5553.719) us 5157.734 us +/-487.979 (min: 4637.888 / max: 6057.088) us

> size = 10000000
> numpy
> select : 182717.449 us +/-3022.580 (min:179389.972 / max:199039.259) us
> cupy
> select : 44139.386 us +/-74.425 (min:44063.589 / max:44533.821) us 47515.994 us +/-71.739 (min:47440.098 / max:47889.217) us

Testing showed that a large time of execution went behind creating those 2 new combined arrays from the smaller individual ones and I couldnt find a way to reasonably reduce that time.

Is there something that I missed or do you have any suggestions ?

emcastillo commented:
Thanks for trying it. 
Let's proceed with it as-is now

emcastillo commented:
Jenkins, test this please
